http/basic/src/test/java/io/cloudevents/http/impl/HttpMessageWriterTest.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2018-Present The CloudEvents Authors
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.cloudevents.http.impl;
+
+import io.cloudevents.CloudEvent;
+import io.cloudevents.SpecVersion;
+import io.cloudevents.core.message.impl.GenericStructuredMessageReader;
+import io.cloudevents.core.mock.CSVFormat;
+import io.cloudevents.types.Time;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.stream.Stream;
+
+import static io.cloudevents.core.test.Data.*;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class HttpMessageWriterTest {
+
+    @ParameterizedTest
+    @MethodSource("io.cloudevents.core.test.Data#allEventsWithoutExtensions")
+    void testReplyWithStructured(CloudEvent event) throws IOException {
+
+        final byte[][] body = {null};
+        final Map<String, String> headers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+        GenericStructuredMessageReader.from(event, CSVFormat.INSTANCE).visit(
+            new HttpMessageWriter(
+                headers::put,
+                b -> {
+                    body[0] = b;
+                }
+            ));
+        assertThat(headers.get("content-type"))
+            .isEqualTo(CSVFormat.INSTANCE.serializedContentType());
+        assertThat(body[0])
+            .isEqualTo(CSVFormat.INSTANCE.serialize(event));
+
+    }
+
+    @ParameterizedTest
+    @MethodSource("binaryTestArguments")
+    void testReplyWithBinary(CloudEvent event, Headers expectedHeaders, byte[] expectedBody) {
+
+        final byte[][] body = {null};
+        final Map<String, String> headers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+
+        new HttpMessageWriter(
+            headers::put,
+            b -> {
+                body[0] = b;
+            }
+        ).writeBinary(event);
+
+        expectedHeaders.forEach((k, v) -> {
+            assertThat(headers.get(k)).isEqualTo(v);
+        });
+        if (expectedBody != null) {
+            assertThat(body[0]).isEqualTo(expectedBody);
+        }
+    }
+
+    public static Stream<Arguments> binaryTestArguments() {
```

Comment:
This code is already copypasted with little changes at multiple places. We should refactor that latter (not in this PR) if possible.

Reply:
Maybe add a todo to the code itself?